src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractWorkspacePullDiagnosticsHandler.cs

```diff
@@ -76,23 +81,50 @@ private void OnLspTextChanged(object? sender, EventArgs e)
 
     private void UpdateLspChanged()
     {
-        Interlocked.Exchange(ref _lspChanged, 1);
+        lock (_gate)
+        {
+            // Loop through our map of source -> has changed and mark them as all having changed.
+            foreach (var category in _categoryToLspChanged.Keys.ToImmutableArray())
+            {
+                _categoryToLspChanged[category] = true;
+            }
+        }
     }
 
-    protected override async Task WaitForChangesAsync(RequestContext context, CancellationToken cancellationToken)
+    protected override async Task WaitForChangesAsync(string? category, RequestContext context, CancellationToken cancellationToken)
     {
+        // A null category counts a separate category and should track changes independently of other categories, so we'll add an empty entry in our map for it.
+        category ??= string.Empty;
+
         // Spin waiting until our LSP change flag has been set.  When the flag is set (meaning LSP has changed),
         // we reset the flag to false and exit out of the loop allowing the request to close.
         // The client will automatically trigger a new request as soon as we close it, bringing us up to date on diagnostics.
-        while (Interlocked.CompareExchange(ref _lspChanged, value: 0, comparand: 1) == 0)
+        while (!HasChanged())
         {
             // There have been no changes between now and when the last request finished - we will hold the connection open while we poll for changes.
             await Task.Delay(TimeSpan.FromMilliseconds(100), cancellationToken).ConfigureAwait(false);
         }
 
-        context.TraceInformation("Closing workspace/diagnostics request");
         // We've hit a change, so we close the current request to allow the client to open a new one.
+        context.TraceInformation("Closing workspace/diagnostics request");
         return;
+
+        bool HasChanged()
```

Comment:
nit: i'd prefer this as a normal method, with 'category' passed in.  but up do you.